### PR TITLE
Fix flaky test_beat.test_xpack Metricbeat system test

### LINKED ISSUE #13947: Flaky test: test_beat.Test.test_xpack
## Flaky Test

* **Test Name:** test_beat.Test.test_xpack
* **Link:** https://github.com/elastic/beats/blob/d1cd81a15b8645ca83e2932d058fb64149d4fe00/metricbeat/tests/system/test_beat.py#L23
* **Branch:** `master`

### Stack Trace

```
08:47:47 ======================================================================
08:47:47 ERROR: beat-xpack module tests
08:47:47 ----------------------------------------------------------------------
08:47:47 Traceback (most recent call last):
08:47:47   File "/go/src/github.com/elastic/beats/metricbeat/tests/system/test_beat.py", line 45, in test_xpack
08:47:47     self.wait_until(lambda: self.output_lines() > 0)
08:47:47   File "/go/src/github.com/elastic/beats/libbeat/tests/system/beat/beat.py", line 352, in wait_until
08:47:47     "Waited {} seconds.".format(max_timeout))
08:47:47 TimeoutError: Timeout waiting for 'cond' to be true. Waited 10 seconds.
```

### PR DESCRIPTION
**Pull request by ycombinator**
Resolves #13947.

This PR:
* fixes the hostnames used to refer to the **monitored** Metricbeat service in the **monitoring** Metricbeat instance's `beat` module configuration.
* removes the naive `sleep` and uses a `wait_until` approach instead.

### CODE REVIEW

File: metricbeat/tests/system/test_beat.py [hunk 1]
```diff
@@ -39,9 +40,13 @@ def test_xpack(self):
...
+        self.wait_until(cond=self.mb_index_exists, max_timeout=60)
 
         proc = self.start_beat()
         self.wait_until(lambda: self.output_lines() > 0)
         proc.check_kill_and_wait()
         self.assert_no_logged_warnings()
+
+    def mb_index_exists(self):
+        es = Elasticsearch([self.get_elasticsearch_url()])
+        return len(es.cat.indices(index='metricbeat-*', h='index')) > 0
```
Comment (ycombinator) [inline metricbeat/tests/system/test_beat.py]: Hey @jsoriano, I could use some guidance here. CI is failing on this line with this error:

```
17:00:38 ======================================================================
17:00:38 ERROR: beat-xpack module tests
17:00:38 ----------------------------------------------------------------------
17:00:38 Traceback (most recent call last):
17:00:38   File "/go/src/github.com/elastic/beats/metricbeat/tests/system/test_beat.py", line 43, in test_xpack
17:00:38     self.wait_until(cond=self.mb_index_exists, max_timeout=60)
17:00:38   File "/go/src/github.com/elastic/beats/libbeat/tests/system/beat/beat.py", line 349, in wait_until
17:00:38     while not cond():
17:00:38   File "/go/src/github.com/elastic/beats/metricbeat/tests/system/test_beat.py", line 52, in mb_index_exists
17:00:38     return len(es.cat.indices(index='metricbeat-*', h='index')) > 0
17:00:38   File "/go/src/github.com/elastic/beats/metricbeat/build/python-env/local/lib/python2.7/site-packages/elasticsearch/client/utils.py", line 76, in _wrapped
17:00:38     return func(*args, params=params, **kwargs)
17:00:38   File "/go/src/github.com/elastic/beats/metricbeat/build/python-env/local/lib/python2.7/site-packages/elasticsearch/client/cat.py", line 137, in indices
17:00:38     'indices', index), params=params)
17:00:38   File "/go/src/github.com/elastic/beats/metricbeat/build/python-env/local/lib/python2.7/site-packages/elasticsearch/transport.py", line 318, in perform_request
17:00:38     status, headers_response, data = connection.perform_request(method, url, params, body, headers=headers, ignore=ignore, timeout=timeout)
17:00:38   File "/go/src/github.com/elastic/beats/metricbeat/build/python-env/local/lib/python2.7/site-packages/elasticsearch/connection/http_urllib3.py", line 181, in perform_request
17:00:38     raise ConnectionError('N/A', str(e), e)
17:00:38 ConnectionError: ConnectionError(<urllib3.connection.HTTPConnection object at 0x7f23baa492d0>: Failed to establish a new connection: [Errno 111] Connection refused) caused by: NewConnectionError(<urllib3.connection.HTTPConnection object at 0x7f23baa492d0>: Failed to establish a new connection: [Errno 111] Connection refused)
...
17:00:38 NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7f23baa492d0>: Failed to establish a new connection: [Errno 111] Connection refused
17:00:38 elasticsearch: DEBUG: > None
17:00:38 elasticsearch: WARNING: GET http://localhost:9200/_cat/indices/metricbeat-*?h=index [status:N/A request:0.001s]
```

I believe the problem is that `self.get_elasticsearch_url()` is returning `http://localhost:9200`, which is not how to address the Elasticsearch service running in the Docker container. Do you know the right way to address it?

Comment (ycombinator) [reply metricbeat/tests/system/test_beat.py]: Never mind, I figured it out:

```python
self.compose_host("elasticsearch")
```

**Comment by ycombinator** [comment]
jenkins, test this

**Comment by ycombinator** [comment]
jenkins, test this

**Comment by ycombinator** [comment]
Looking at the current Travis failure it doesn't look like this PR removes the flakiness from the test. Will keep working on it.

File: metricbeat/tests/system/test_beat.py [hunk 2]
```diff
@@ -17,31 +18,44 @@ def test_metricsets(self, metricset):
...
         proc.check_kill_and_wait()
         self.assert_no_logged_warnings()
+
+    def delete_mb_indices(self):
+        es = Elasticsearch([self.get_elasticsearch_url()])
+        return es.indices.delete("metricbeat-*")
+
+    def mb_index_exists(self):
+        es = Elasticsearch([self.get_elasticsearch_url()])
+        return len(es.cat.indices(index='metricbeat-*')) > 0
```
Comment (jsoriano) [inline metricbeat/tests/system/test_beat.py]: I guess this approach can affect or be affected by other tests if run in parallel, if we need to check for the presence of an specific index we should probably use a unique index name randomly generated for this test. We might set a custom `setup.ilm.rollover_alias` for that.

Comment (ycombinator) [reply metricbeat/tests/system/test_beat.py]: So really what I want to check here is whether the _monitored_ Metricbeat instance has been able to connect to Elasticsearch. I was using the presence of the index as that check but I could use something else too. Ideally I would be able to look for the presence of a line in the monitored Metricbeat's log.

**Comment by ycombinator** [comment]
@jsoriano This PR is ready for re-review. The test now waits until the monitored Metricbeat has connected to Elasticsearch.

File: metricbeat/tests/system/test_beat.py (hunk 2 above)
Comment (ycombinator) [reply metricbeat/tests/system/test_beat.py]: Implemented this check now.

File: libbeat/tests/system/beat/compose.py [hunk 3]
```diff
@@ -215,3 +215,17 @@ def find_compose_path(cls):
...
             if current == '':  # We have reached root
                 raise Exception("failed to find a docker-compose.yml file")
+
+    @classmethod
+    def get_service_log(cls, service):
+        container = cls.compose_project().containers(service_names=[service])[0]
+        return container.logs()
+
+    @classmethod
+    def service_log_contains(cls, service, msg):
```
Comment (ycombinator) [inline libbeat/tests/system/beat/compose.py]: I have deliberately chosen to keep the implementation of this function simple — without regex matching or case ignoring. We can add those features as we need them, IMO.

**Pull request merged**